crates/kumod/src/smtp_server.rs

```diff
@@ -2764,3 +2825,50 @@ mod test {
         ));
     }
 }
+
+async fn extract_proxy_protocol_peer_address<S: AsyncReadExt + Unpin>(
+    socket: &mut S,
+    _orig_peer_address: SocketAddr,
+) -> anyhow::Result<SocketAddr> {
+    const PROXY_V2_MINIMUM_LENGTH: usize = 16;
+    let mut header_buf = [0u8; PROXY_V2_MINIMUM_LENGTH];
+    socket.read_exact(&mut header_buf).await
+        .map_err(|e| anyhow::anyhow!("Failed to read PROXY v2 preamble: {e:#}"))?;
+
+    let len = u16::from_be_bytes([header_buf[14], header_buf[15]]) as usize;
+    let mut rest = vec![0u8; len];
+    socket.read_exact(&mut rest).await
+        .map_err(|e| anyhow::anyhow!("Failed to read the rest of the PROXY v2 header: {e:#}"))?;
+
+    let mut full_header = header_buf.to_vec();
+    full_header.extend_from_slice(&rest);
+
+    match ppp::v2::Header::try_from(full_header.as_slice()) {
+        Ok(header) => match header.addresses {
+            ppp::v2::Addresses::IPv4(addr) => Ok(SocketAddr::new(
+                std::net::IpAddr::V4(addr.source_address),
+                addr.source_port,
+            )),
+            ppp::v2::Addresses::IPv6(addr) => Ok(SocketAddr::new(
+                std::net::IpAddr::V6(addr.source_address),
+                addr.source_port,
+            )),
+            _ => Err(anyhow::anyhow!("Unsupported PROXY protocol address family")),
+        },
+        Err(e) => Err(anyhow::anyhow!("Failed to parse proxy protocol header: {e:#}")),
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum ProxyMode {
+    None,
```

Comment:
Let's remove the `None` variant from this, as we seem to be using `Option<ProxyMode>` throughout, and `Some(ProxyMode::None)` is redundant/ambiguous wrt. to `None`.

That will work better with using a list of allowed proxy modes; it would be better to have an empty list to indicate `no-allowed-proxy-modes` than a list containing `ProxyMode::None`, especially if that list could then contain some other proxy mode alongside it!

```suggestion
```